File: sdk/python/packages/flet/src/flet/core/search_bar.py
import time
from dataclasses import field
from typing import List, Optional

from flet.core.border import BorderSide
from flet.core.box import BoxConstraints
from flet.core.buttons import OutlinedBorder
from flet.core.constrained_control import ConstrainedControl
from flet.core.control import Control, control
from flet.core.padding import OptionalPaddingValue, PaddingValue
from flet.core.text_style import TextStyle
from flet.core.textfield import KeyboardType, TextCapitalization
from flet.core.types import (
    ColorValue,
    ControlStateValue,
    OptionalColorValue,
    OptionalControlEventCallable,
    OptionalNumber,
)

__all__ = ["SearchBar"]


@control("SearchBar")
class SearchBar(ConstrainedControl):
    """
    Manages a "search view" route that allows the user to select one of the suggested completions for a search query.

    -----

    Online docs: https://flet.dev/docs/controls/searchbar
    """

    controls: List[Control] = field(default_factory=list)
    value: Optional[str] = None
    bar_leading: Optional[Control] = None
    bar_trailing: Optional[List[Control]] = None
    bar_hint_text: Optional[str] = None
    bar_bgcolor: ControlStateValue[ColorValue] = None
    bar_overlay_color: ControlStateValue[ColorValue] = None
    bar_shadow_color: ControlStateValue[ColorValue] = None
    bar_surface_tint_color: ControlStateValue[ColorValue] = None
    bar_elevation: ControlStateValue[OptionalNumber] = None
    bar_border_side: ControlStateValue[BorderSide] = None
    bar_shape: ControlStateValue[OutlinedBorder] = None
    bar_text_style: ControlStateValue[TextStyle] = None
    bar_hint_text_style: ControlStateValue[TextStyle] = None
    bar_padding: ControlStateValue[PaddingValue] = None
    bar_scroll_padding: OptionalPaddingValue = None
    view_leading: Optional[Control] = None
    view_trailing: Optional[List[Control]] = None
    view_elevation: OptionalNumber = None
    view_bgcolor: OptionalColorValue = None
    view_hint_text: Optional[str] = None
    view_side: Optional[BorderSide] = None
    view_shape: Optional[OutlinedBorder] = None
    view_header_text_style: Optional[TextStyle] = None
    view_hint_text_style: Optional[TextStyle] = None
    view_size_constraints: Optional[BoxConstraints] = None
    view_header_height: OptionalNumber = None
    divider_color: OptionalColorValue = None
    capitalization: Optional[TextCapitalization] = None
    full_screen: Optional[bool] = None
    keyboard_type: Optional[KeyboardType] = None
    view_surface_tint_color: OptionalColorValue = None
    autofocus: Optional[bool] = None
    on_tap: OptionalControlEventCallable = None
    on_tap_outside_bar: OptionalControlEventCallable = None
    on_submit: OptionalControlEventCallable = None
    on_change: OptionalControlEventCallable = None
    on_focus: OptionalControlEventCallable = None
    on_blur: OptionalControlEventCallable = None

    def __contains__(self, item):
        return item in self.controls

    def before_update(self):
        super().before_update()
        # self._set_attr_json("barBgcolor", self.__bar_bgcolor, wrap_attr_dict=True)
        # self._set_attr_json(
        #     "barOverlayColor", self.__bar_overlay_color, wrap_attr_dict=True
        # )
        # self._set_attr_json(
        #     "barHintTextStyle", self.__bar_hint_text_style, wrap_attr_dict=True
        # )
        # self._set_attr_json(
        #     "barSurfaceTintColor", self.__bar_surface_tint_color, wrap_attr_dict=True
        # )
        # self._set_attr_json("barElevation", self.__bar_elevation, wrap_attr_dict=True)
        # self._set_attr_json(
        #     "barBorderSide", self.__bar_border_side, wrap_attr_dict=True
        # )
        # self._set_attr_json("barShape", self.__bar_shape, wrap_attr_dict=True)
        # self._set_attr_json("barTextStyle", self.__bar_text_style, wrap_attr_dict=True)
        # self._set_attr_json("barPadding", self.__bar_padding, wrap_attr_dict=True)
        # self._set_attr_json(
        #     "barShadowColor", self.__bar_shadow_color, wrap_attr_dict=True
        # )

    # Public methods
    def focus(self):
        self._set_attr_json("focus", str(time.time()))
        self.update()

    def blur(self):
        self._set_attr_json("blur", str(time.time()))
        self.update()

    def open_view(self):
        m = {
            "n": "openView",
            "i": str(time.time()),
            "p": {},
        }
        self._set_attr_json("method", m)
        self.update()

    def close_view(self, text: str = ""):
        m = {
            "n": "closeView",
            "i": str(time.time()),
            "p": {"text": text},
        }
        self.value = text
        self._set_attr_json("method", m)
        self.update()

File: packages/flet/lib/src/controls/search_bar.dart
import 'package:flutter/material.dart';

import '../extensions/control.dart';
import '../models/control.dart';
import '../utils/borders.dart';
import '../utils/box.dart';
import '../utils/colors.dart';
import '../utils/edge_insets.dart';
import '../utils/form_field.dart';
import '../utils/numbers.dart';
import '../utils/text.dart';
import 'base_controls.dart';

class SearchBarControl extends StatefulWidget {
  final Control control;

  const SearchBarControl({super.key, required this.control});

  @override
  State<SearchBarControl> createState() => _SearchBarControlState();
}

class _SearchBarControlState extends State<SearchBarControl> {
  late final SearchController _controller;
  bool _focused = false;
  TextCapitalization _textCapitalization = TextCapitalization.none;
  late final FocusNode _focusNode;
  String? _lastFocusValue;
  String? _lastBlurValue;

  @override
  void initState() {
    super.initState();
    _controller = SearchController();
    _controller.addListener(_searchTextChanged);
    _focusNode = FocusNode();
    _focusNode.addListener(_onFocusChange);
    widget.control.addInvokeMethodListener(_invokeMethod);
  }

  void _onFocusChange() {
    setState(() {
      _focused = _focusNode.hasFocus;
    });
    widget.control.triggerEvent(_focusNode.hasFocus ? "focus" : "blur");
  }

  @override
  void dispose() {
    _controller.removeListener(_searchTextChanged);
    _controller.dispose();
    _focusNode.removeListener(_onFocusChange);
    _focusNode.dispose();
    widget.control.removeInvokeMethodListener(_invokeMethod);
    super.dispose();
  }

  void _searchTextChanged() {
    _textCapitalization = parseTextCapitalization(
        widget.control.getString("capitalization"), TextCapitalization.none)!;
    _updateValue(_controller.text);
  }

  void _updateValue(String value) {
    value = applyCapitalization(value);
    if (_controller.value.text != value) {
      _controller.value = TextEditingValue(
        text: value,
        selection: TextSelection.collapsed(offset: value.length),
      );
    }

    widget.control.updateProperties({"value": value});
  }

  Future<dynamic> _invokeMethod(String name, dynamic args) async {
    debugPrint("SearchBar.$name($args)");
    switch (name) {
      case "close_view":
        if (_controller.isOpen) {
          var text = args["text"];
          _updateValue(text);
          _controller.closeView(text);
        }
      case "open_view":
        if (!_controller.isOpen) {
          _controller.openView();
        }
      case "focus":
        _focusNode.requestFocus();
      case "blur":
        // todo: test this method
        _focusNode.unfocus(
            disposition: UnfocusDisposition.previouslyFocusedChild);
      default:
        throw Exception("Unknown SearchBar method: $name");
    }
  }

  String applyCapitalization(String text) {
    switch (_textCapitalization) {
      /// Capitalizes the first character of each word.
      case TextCapitalization.words:
        return text
            .split(RegExp(r'\s+'))
            .map((word) => word.isNotEmpty
                ? word[0].toUpperCase() + word.substring(1).toLowerCase()
                : word)
            .join(' ');

      /// Capitalizes the first character of each sentence.
      case TextCapitalization.sentences:
        return text
            .split('. ')
            .map((sentence) => sentence.isNotEmpty
                ? sentence.trimLeft()[0].toUpperCase() +
                    sentence.substring(1).toLowerCase()
                : sentence)
            .join('. ');

      /// Capitalizes all characters.
      case TextCapitalization.characters:
        return text.toUpperCase();

      /// No change.
      case TextCapitalization.none:
        return text;
    }
  }

  @override
  Widget build(BuildContext context) {
    debugPrint("SearchAnchor build: ${widget.control.id}");

    var value = widget.control.getString("value", "")!;
    if (value != _controller.text) {
      WidgetsBinding.instance.addPostFrameCallback((_) {
        _controller.text = value;
      });
    }

    bool onChange = widget.control.getBool("on_change", false)!;
    bool onTap = widget.control.getBool("on_tap", false)!;
    bool onSubmit = widget.control.getBool("on_submit", false)!;
    TextInputType keyboardType =
        widget.control.getTextInputType("keyboard_type", TextInputType.text)!;

    var focusValue = widget.control.getString("focus");
    var blurValue = widget.control.getString("blur");
    if (focusValue != null && focusValue != _lastFocusValue) {
      _lastFocusValue = focusValue;
      _focusNode.requestFocus();
    }
    if (blurValue != null && blurValue != _lastBlurValue) {
      _lastBlurValue = blurValue;
      _focusNode.unfocus();
    }

    var theme = Theme.of(context);

    Widget anchor = SearchAnchor(
        searchController: _controller,
        headerHintStyle:
            widget.control.getTextStyle("view_hint_text_style", theme),
        headerTextStyle:
            widget.control.getTextStyle("view_header_text_style", theme),
        viewSide: widget.control.getBorderSide("view_side", theme),
        isFullScreen: widget.control.getBool("full_screen", false),
        viewBackgroundColor: widget.control.getColor("view_bgcolor", context),
        dividerColor: widget.control.getColor("divider_color", context),
        viewHintText: widget.control.getString("view_hint_text"),
        viewElevation: widget.control.getDouble("view_elevation"),
        headerHeight: widget.control.getDouble("view_header_height"),
        viewConstraints:
            widget.control.getBoxConstraints("view_size_constraints"),
        viewShape: widget.control.getShape("view_shape", theme),
        viewTrailing: widget.control.buildWidgets("view_trailing"),
        viewLeading: widget.control.buildWidget("view_leading"),
        viewOnSubmitted: onSubmit
            ? (String value) {
                _updateValue(value);
                widget.control.triggerEvent("submit", value);
              }
            : null,
        viewOnChanged: onChange
            ? (String value) {
                _updateValue(value);
                widget.control.triggerEvent("change", value);
              }
            : null,
        viewSurfaceTintColor:
            widget.control.getColor("view_surface_tint_color", context),
        textCapitalization: _textCapitalization,
        keyboardType: keyboardType,
        builder: (BuildContext context, SearchController controller) {
          return SearchBar(
            controller: controller,
            keyboardType: keyboardType,
            textCapitalization: _textCapitalization,
            autoFocus: widget.control.getBool("autofocus", false)!,
            focusNode: _focusNode,
            hintText: widget.control.getString("bar_hint_text"),
            elevation: widget.control.getWidgetStateDouble("bar_elevation"),
            shape:
                widget.control.getWidgetStateOutlinedBorder("bar_shape", theme),
            padding: widget.control.getWidgetStatePadding("bar_padding"),
            textStyle:
                widget.control.getWidgetStateTextStyle("bar_text_style", theme),
            hintStyle: widget.control
                .getWidgetStateTextStyle("bar_hint_text_style", theme),
            shadowColor:
                widget.control.getWidgetStateColor("bar_shadow_color", theme),
            surfaceTintColor: widget.control
                .getWidgetStateColor("bar_surface_tint_color", theme),
            side: widget.control
                .getWidgetStateBorderSide("bar_border_side", theme),
            backgroundColor:
                widget.control.getWidgetStateColor("bar_bgcolor", theme),
            overlayColor:
                widget.control.getWidgetStateColor("bar_overlay_color", theme),
            scrollPadding: widget.control
                .getPadding("bar_scroll_padding", const EdgeInsets.all(20.0))!,
            leading: widget.control.buildWidget("bar_leading"),
            trailing: widget.control.buildWidgets("bar_trailing"),
            onTap: onTap ? () => widget.control.triggerEvent("tap") : null,
            onTapOutside: widget.control.getBool("on_tap_outside_bar", false)!
                ? (PointerDownEvent? event) =>
                    widget.control.triggerEvent("tap_outside_bar")
                : null,
            onSubmitted: onSubmit
                ? (String value) {
                    _updateValue(value);
                    widget.control.triggerEvent("submit", value);
                  }
                : null,
            onChanged: onChange
                ? (String value) {
                    _updateValue(value);
                    widget.control.triggerEvent("change", value);
                  }
                : null,
          );
        },
        suggestionsBuilder:
            (BuildContext context, SearchController controller) {
          return widget.control.buildWidgets("controls");
        });

    return ConstrainedControl(control: widget.control, child: anchor);
  }
}
